## <b> problem statement </b>
- <b> <i> FoodHunter </i> </b> , a leading food delivery app, has experienced a drop in revenue over the months. The goal is to investigate the intensity of the decline, identify the causes and contributing factors behind this decline and help to develop strategies to overcome these challenges.

## <b> loading the SQL extension and connecting to the <i> foodhunter database </i> </b>

In [1]:
%load_ext sql
%sql mysql+mysqlconnector://root:root@localhost/

In [2]:
%%sql
use foodhunter;
show tables;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
6 rows affected.


[('customers',),
 ('drivers',),
 ('food_items',),
 ('orders',),
 ('orders_items',),
 ('restaurants',)]

## <b> Investigating tables</b>

### <b> orders</b>

In [3]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "orders";

 * mysql+mysqlconnector://root:***@localhost/
13 rows affected.


[('customer_id',),
 ('delivered_date',),
 ('delivered_time',),
 ('delivery_address',),
 ('delivery_fee',),
 ('discount',),
 ('driver_id',),
 ('final_price',),
 ('order_date',),
 ('order_id',),
 ('order_rating',),
 ('order_time',),
 ('total_price',)]

In [4]:
%%sql
#understanding datatype, field characteristics
desc foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/
13 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('customer_id', 'bigint', 'YES', '', None, ''),
 ('order_date', 'datetime', 'YES', '', None, ''),
 ('order_time', 'time', 'YES', '', None, ''),
 ('delivered_date', 'text', 'YES', '', None, ''),
 ('delivered_time', 'time', 'YES', '', None, ''),
 ('delivery_address', 'text', 'YES', '', None, ''),
 ('driver_id', 'bigint', 'YES', '', None, ''),
 ('total_price', 'bigint', 'YES', '', None, ''),
 ('delivery_fee', 'double', 'YES', '', None, ''),
 ('discount', 'double', 'YES', '', None, ''),
 ('final_price', 'double', 'YES', '', None, ''),
 ('order_rating', 'double', 'YES', '', None, '')]

In [5]:
%%sql

# identifying the timeframe for our root-cause analysis
select min(order_date) as first_date_orders, 
max(order_date) as last_date_orders 
from foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/
1 rows affected.


[(datetime.datetime(2022, 6, 1, 0, 0), datetime.datetime(2022, 9, 29, 0, 0))]

### observations
* We've a data from 1st June 2022 to 29th September 2022 to understand why there was a drop in revenue for foodhunter during this timeframe and to identify the underlying causes.

In [6]:
%%sql

#Investigating basic structure of data...
select
*
from
foodhunter.orders
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(1, 7420, datetime.datetime(2022, 6, 1, 0, 0), datetime.timedelta(seconds=180), '2022-06-01', datetime.timedelta(seconds=960), '7 Cedar Lane', 93, 150, 0.0, 30.0, 120.0, 3.5554637033425025),
 (2, 1895, datetime.datetime(2022, 6, 1, 0, 0), datetime.timedelta(seconds=240), '2022-06-01', datetime.timedelta(seconds=1080), '7 Lexington Place', 138, 9, 5.0, 0.0, 14.0, 5.0),
 (3, 6651, datetime.datetime(2022, 6, 1, 0, 0), datetime.timedelta(seconds=300), '2022-06-01', datetime.timedelta(seconds=1200), '20 Linden Lane', 187, 117, 0.0, 23.4, 93.6, 3.393488878617012),
 (4, 9032, datetime.datetime(2022, 6, 1, 0, 0), datetime.timedelta(seconds=300), '2022-06-01', datetime.timedelta(seconds=1440), '6 Rosedale Avenue', 10, 105, 0.0, 21.0, 84.0, 5.0),
 (5, 2479, datetime.datetime(2022, 6, 1, 0, 0), datetime.timedelta(seconds=360), '2022-06-01', datetime.timedelta(seconds=2040), '36 Cherrywood Lane', 219, 27, 2.5, 5.4, 24.1, 3.7930482971558623)]

### <b> orders_items</b>

In [7]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "orders_items";

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('item_id',), ('order_id',), ('quantity',)]

In [8]:
%%sql
desc foodhunter.orders_items;

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('item_id', 'int', 'YES', '', None, ''),
 ('quantity', 'int', 'YES', '', None, '')]

In [9]:
%%sql

select
*
from
foodhunter.orders_items
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(1, 176, 3), (2, 108, 1), (3, 339, 3), (4, 380, 1), (4, 108, 1)]

### <b> restaurants</b>

In [10]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "restaurants";

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[('address',), ('cuisine',), ('restaurant_id',), ('restaurant_name',)]

In [11]:
%%sql
desc foodhunter.restaurants;

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[('restaurant_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_name', 'text', 'YES', '', None, ''),
 ('cuisine', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

In [12]:
%%sql
select
*
from
foodhunter.restaurants
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(1, "L'Artusi", 'Italian', '"3540 7th Ave'),
 (2, 'Loring Place', 'Chinese', '"6947 4th Ave'),
 (3, 'Café China', 'Chinese', '"9675 7th Ave'),
 (4, 'Minetta Tavern', 'Thai', '"609 2nd Ave'),
 (5, 'Carbone', 'Chinese', '"871 8th Ave')]

### <b> customers</b>

In [13]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "customers";

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[('address',),
 ('customer_id',),
 ('first_name',),
 ('last_name',),
 ('phone_number',)]

In [14]:
%%sql
desc foodhunter.customers;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[('customer_id', 'int', 'NO', 'PRI', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('phone_number', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

In [15]:
%%sql
select
*
from
foodhunter.customers
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(1, 'Donald', 'Arend', '9020474801', '49 Madison Avenue'),
 (2, 'Richard', 'Krick', '7040231632', '32 Vineyard Drive'),
 (3, 'Pauline', 'Hardman', '9180289990', '18 Elmwood Place'),
 (4, 'Lawrence', 'Sultzer', '8278748624', '77 Rosedale Avenue'),
 (5, 'Michael', 'Honeycutt', '8821565245', '56 Broad Street')]

### <b> drivers</b>

In [16]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "drivers";

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('driver_id',), ('name',), ('rating',)]

In [17]:
%%sql
desc foodhunter.drivers;

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('driver_id', 'int', 'NO', 'PRI', None, ''),
 ('name', 'text', 'YES', '', None, ''),
 ('rating', 'double', 'YES', '', None, '')]

In [18]:
%%sql
select
*
from
foodhunter.drivers
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(1, 'Avery Yost', 3.1),
 (2, 'Anthony Pegram', 4.3),
 (3, 'Elaine Kimes', 4.2),
 (4, 'Julie West', 3.7),
 (5, 'Jamie Hyman', 4.3)]

### <b> food_items</b>

In [19]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "food_items";

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('calories',),
 ('food_type',),
 ('item_id',),
 ('item_name',),
 ('price',),
 ('restaurant_id',)]

In [20]:
%%sql
desc foodhunter.food_items;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('item_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_id', 'int', 'YES', 'MUL', None, ''),
 ('item_name', 'text', 'YES', '', None, ''),
 ('food_type', 'text', 'YES', '', None, ''),
 ('price', 'text', 'YES', '', None, ''),
 ('calories', 'double', 'YES', '', None, '')]

In [21]:
%%sql
select
*
from
foodhunter.food_items
limit 5;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[(101, 47, 'Coconut Ice Cream', 'veg', None, 500.0),
 (102, 93, 'Vegetable Pakoras', 'veg', None, 1000.0),
 (103, 24, 'Wonton soup', 'non-veg', None, 400.0),
 (104, 23, 'Spring rolls', 'non-veg', None, 500.0),
 (105, 77, 'Chinotto', 'veg', None, 500.0)]

### <b> The rate of change in order count and revenue over the months </b>

In [22]:
%%sql

with order_revenue_summary as (
  select 
  concat(year(order_date), '-', month(order_date)) as year__month, 
  count(distinct order_id) as ordercount, 
  sum(final_price) as tot_revenue 
  from foodhunter.orders 
  group by concat(year(order_date), '-', month(order_date))
),
lag_order_revenue_summary as(
  select
  year__month,
  ordercount,
  lag(ordercount) over(order by year__month asc) as lag_order_count,
  lead(ordercount, 3) over(order by year__month asc) as lead_order_count,
  tot_revenue,
  lag(tot_revenue) over(order by year__month asc) as lag_tot_revenue,
  lead(tot_revenue, 3) over(order by year__month asc) as lead_tot_revenue
  from order_revenue_summary
)

select 
  year__month,
  ordercount,
  lag_order_count,
  round(((ordercount - lag_order_count) / lag_order_count) * 100, 2) as orderCountChangeRate,
  round(((lead_order_count - ordercount) / ordercount)*100, 2) as overallOrderCountChangeRate,
  round(tot_revenue, 2) as tot_revenue,
  round(lag_tot_revenue, 2) as lag_tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue) * 100, 2) as totRevenueChangeRate,
  round(((lead_tot_revenue - tot_revenue) / tot_revenue)*100, 2) as overallTotRevenueChangeRate
from lag_order_revenue_summary;

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[('2022-6', 12502, None, None, Decimal('-25.09'), 347577.5, None, None, -25.73),
 ('2022-7', 11144, 12502, Decimal('-10.86'), None, 308601.5, 347577.5, -11.21, None),
 ('2022-8', 10107, 11144, Decimal('-9.31'), None, 283365.9, 308601.5, -8.18, None),
 ('2022-9', 9365, 10107, Decimal('-7.34'), None, 258161.1, 283365.9, -8.89, None)]

#### <b> observations </b>
* From the data, we observe a downward trend in order count with <b>25% decline</b> and total revenue with <b>26% decline</b> from June 2022 to September 2022.
  
* <b> The order count </b> shows a 10.86% drop from June to July, 9.31% drop from July to August and 7.34% drop from August to September. While the rate of decline is slowing, the overall drop in order count is still significant.

* On the Otherhand, <b> The total revenue </b> shows an 11.21% drop from June to July, 8.18% drop from July to August and 8.89% drop from August to September. Although the rate of decline fluctuates, the overall decrease in total revenue is significant and requires immediate attention.

* Based on the rate of change, we can infer that the factors causing the decline are fading out. <i><b>[Assumption] :</b> It seems that recovering phase gets started or marketing strategies have started to adjust and adapt, leading to a reduction in the impact of losses.</i>

* To confirm this, It is important to analyze the data at various granularities, as revenue shows signs of volatility <i>[out of control] </i> in its rate of change.

### <b> Weekday vs Weekend: Monthly Variations in Orders and Revenue </b>

In [23]:
%%sql

select
  day_category,
  count(distinct order_id) as ordercount,
  round(sum(final_price), 2) as tot_revenue,
  case when day_category = "Week-Day" 
    then count(distinct order_id) / 5
    else count(distinct order_id) / 2
  end as ordercount_per_dayCat_approx,
  case when day_category = "Week-Day"
    then round(sum(final_price), 2) / 5
    else round(sum(final_price), 2) / 2
  end as tot_revenue_per_dayCat_approx
from
  (select
    *,
    case 
     when dayofweek(order_date) in (1,7) then "Week-End"
     else "Week-Day"
    end as day_category
  from foodhunter.orders) q1
group by day_category;

 * mysql+mysqlconnector://root:***@localhost/
2 rows affected.


[('Week-Day', 31412, 870423.1, Decimal('6282.4000'), 174084.62),
 ('Week-End', 11706, 327282.9, Decimal('5853.0000'), 163641.45)]

#### <b> observations </b>
* From this we just confirming, Weekdays are performing well as expected with higher revenue and order count compared to weekends.

In [24]:
%%sql

with weekpart_order_revenue_summary as(
 select
   concat(year(order_date), '-', month(order_date)) as year__month,
   case
     when dayofweek(order_date) in (1,7) then "Week-End"
       else "Week-Day"
   end as day_category,
   count(distinct order_id) as ordercount, 
   sum(final_price) as tot_revenue 
 from foodhunter.orders
   group by concat(year(order_date), '-', month(order_date)), day_category
),
lag_weekpart_summary as(
  select
    year__month,
    day_category,
    ordercount,
    lag(ordercount) over(partition by day_category order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by day_category order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by day_category order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by day_category order by year__month) as lead_overall_rev
  from weekpart_order_revenue_summary
)

select 
  year__month,
  day_category,
  round(ordercount, 2) as ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount)/ ordercount)*100, 2) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue)/ tot_revenue)*100, 2) as totRev_overall_change_in_percent
from lag_weekpart_summary;

 * mysql+mysqlconnector://root:***@localhost/
8 rows affected.


[('2022-6', 'Week-Day', 9043, None, Decimal('-23.82'), 250657.7, None, -24.52),
 ('2022-7', 'Week-Day', 8096, Decimal('-10.47'), None, 224993.4, -10.24, None),
 ('2022-8', 'Week-Day', 7384, Decimal('-8.79'), None, 205586.2, -8.63, None),
 ('2022-9', 'Week-Day', 6889, Decimal('-6.70'), None, 189185.8, -7.98, None),
 ('2022-6', 'Week-End', 3459, None, Decimal('-28.42'), 96919.8, None, -28.83),
 ('2022-7', 'Week-End', 3048, Decimal('-11.88'), None, 83608.1, -13.73, None),
 ('2022-8', 'Week-End', 2723, Decimal('-10.66'), None, 77779.7, -6.97, None),
 ('2022-9', 'Week-End', 2476, Decimal('-9.07'), None, 68975.3, -11.32, None)]

#### <b> observations </b>
* For Weekday, ordercounts shows a downward trend with 24% & total revenue shows a downward trend with 25% from June to September, with the rate of decline gradually slowing over the months.
  
* For weekend, ordercounts shows a downward trend with 28% & total revenue shows a downward trend with 29% over the months, with a fluctuations in the rate of decline.

* From this observation at this level of granularity, we can conclude that revenue changes across the four months for weekends are highly volatile (with a fluctuations in the rate of change), while the rate of decline is high compared to weekday metrics.

* FoodHunter should focus more on weekend operations as a primary focus than weekday operations due to revenue fluctuations (situation is not predictable /<i>out of control</i>/), as the current strategy does not seem effective for weekends in terms of revenue.

### <b> OrderCount & Total Revenue Variation Across Time Segments </b>

#### <b> <i> Segmenting time periods according to typical eating rhythm </i></b>
- #### <b> <i> order_timings => dining_segment </i></b>
- 07:00 – 10:00	=> Breakfast
- 10:01 – 11:59	=> Brunch
- 12:00 – 14:59	=> Lunch
- 15:00 – 18:59	=> Snacks
- 19:00 – 22:00	=> Dinner
- 22:01 – 06:59	=> After hours

In [25]:
%%sql

create or replace view foodhunter_reports.dining_segment_summary as
select
  year__month,
  day_category,
  dining_segment,
  count(distinct order_id) as ordercount,
  sum(final_price) as tot_revenue
from
  (select
     concat(year(order_date), "-", month(order_date)) as year__month,
     case
        when dayofweek(order_date) in (1,7) then "Week-End"
        else "Week-Day"
     end as day_category,
     case
        when order_time between '07:00:00' and '10:00:00' 
        then '1-Breakfast'
        when order_time between '10:01:00' and '11:59:59' 
        then '2-Brunch'
        when order_time between '12:00:00' and '14:59:59' 
        then '3-Lunch'
        when order_time between '15:00:00' and '18:59:59' 
        then '4-Snacks'
        when order_time between '19:00:00' and '22:00:00' 
        then '5-Dinner'
        when (order_time between '22:01:00' and '23:59:59') 
             or 
             (order_time between 
              '00:00:00' and '06:59:59')  
        then '6-After hours'
     end as dining_segment,
     order_id,
     final_price
  from foodhunter.orders) q1
group by year__month, day_category, dining_segment;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.


[]

In [26]:
%%sql
#overall revenue and ordercount shares for dining segments.
select
  dining_segment,
  sum(ordercount) as ordercount,
  round((sum(ordercount) / sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following))*100, 2) as ordercount_share,
  round(sum(tot_revenue), 2) as tot_revenue,
  round((sum(tot_revenue) / sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following))*100, 2) as total_rev_share
from foodhunter_reports.dining_segment_summary
group by dining_segment
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('13726'), Decimal('31.83'), 380464.4, 31.77),
 ('4-Snacks', Decimal('7797'), Decimal('18.08'), 215378.5, 17.98),
 ('3-Lunch', Decimal('5950'), Decimal('13.80'), 167552.6, 13.99),
 ('1-Breakfast', Decimal('5990'), Decimal('13.89'), 166993.6, 13.94),
 ('5-Dinner', Decimal('5890'), Decimal('13.66'), 162552.1, 13.57),
 ('2-Brunch', Decimal('3765'), Decimal('8.73'), 104764.8, 8.75)]

#### <b> observations </b>
* The After Hours segment (10 PM to 7 AM) records the highest number of orders and total revenue, which could be attributed by several factors Limited restaurant availability, restricted mobility, Late-night cravings, social events and irregular sleep schedules also contribute to higher demand.
  
* Brunch ranks the lowest in both order count and total revenue possibly due to it's not aligning with the typical eating habits of most people.

* To prevent delays and resource shortages, delivery partners should be strategically allocated based on the proportion of metrics for the dining segments above.

In [27]:
%%sql

with weekdaylag_dining_segment_summary as (
  select 
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by dining_segment order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by dining_segment order by year__month) as lead_overall_rev
  from  foodhunter_reports.dining_segment_summary 
  where day_category = 'Week-Day')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount) / ordercount)*100, 2) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue) / tot_revenue)*100, 2) as totRev_overall_change_in_percent
from weekdaylag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/
24 rows affected.


[('2022-6', '1-Breakfast', 1215, None, Decimal('-17.86'), 33230.7, None, -18.02),
 ('2022-7', '1-Breakfast', 1144, Decimal('-5.84'), None, 33229.6, -0.0, None),
 ('2022-8', '1-Breakfast', 1018, Decimal('-11.01'), None, 27481.0, -17.3, None),
 ('2022-9', '1-Breakfast', 998, Decimal('-1.96'), None, 27242.2, -0.87, None),
 ('2022-6', '2-Brunch', 806, None, Decimal('-28.78'), 22517.5, None, -30.34),
 ('2022-7', '2-Brunch', 690, Decimal('-14.39'), None, 19289.9, -14.33, None),
 ('2022-8', '2-Brunch', 650, Decimal('-5.80'), None, 17782.7, -7.81, None),
 ('2022-9', '2-Brunch', 574, Decimal('-11.69'), None, 15686.4, -11.79, None),
 ('2022-6', '3-Lunch', 1228, None, Decimal('-20.77'), 33855.7, None, -19.84),
 ('2022-7', '3-Lunch', 1129, Decimal('-8.06'), None, 31745.8, -6.23, None),
 ('2022-8', '3-Lunch', 996, Decimal('-11.78'), None, 28130.3, -11.39, None),
 ('2022-9', '3-Lunch', 973, Decimal('-2.31'), None, 27138.9, -3.52, None),
 ('2022-6', '4-Snacks', 1656, None, Decimal('-29.17'), 46596.2, None, -30.65),
 ('2022-7', '4-Snacks', 1500, Decimal('-9.42'), None, 41610.6, -10.7, None),
 ('2022-8', '4-Snacks', 1380, Decimal('-8.00'), None, 38004.8, -8.67, None),
 ('2022-9', '4-Snacks', 1173, Decimal('-15.00'), None, 32313.7, -14.97, None),
 ('2022-6', '5-Dinner', 1282, None, Decimal('-27.85'), 35413.5, None, -27.38),
 ('2022-7', '5-Dinner', 1088, Decimal('-15.13'), None, 29802.7, -15.84, None),
 ('2022-8', '5-Dinner', 992, Decimal('-8.82'), None, 28083.4, -5.77, None),
 ('2022-9', '5-Dinner', 925, Decimal('-6.75'), None, 25717.6, -8.42, None),
 ('2022-6', '6-After hours', 2856, None, Decimal('-21.36'), 79044.1, None, -22.72),
 ('2022-7', '6-After hours', 2545, Decimal('-10.89'), None, 69314.8, -12.31, None),
 ('2022-8', '6-After hours', 2348, Decimal('-7.74'), None, 66104.0, -4.63, None),
 ('2022-9', '6-After hours', 2246, Decimal('-4.34'), None, 61087.0, -7.59, None)]

In [28]:
%%sql

#revenue and ordercount shares for dining segments for weekday.
with weekday_priority as(
select
a.dining_segment,
a.ordercount,
a.ordercount_share,
q2.overallChange_ordercount,
a.tot_revenue,
a.tot_revenue_share,
q2.overallChange_totRev,
#just to set priority on framing out recovery plan based on severity
(-(1.5*a.tot_revenue_share) * (q2.overallChange_totRev)) as attentionScore, #providing 0.5 units more priority for revenue share from total revenue decline
dense_rank() over(order by tot_revenue desc) as rank_Rev, 
dense_rank() over(order by overallChange_totRev asc) as rank_Overall_Drop_Rev,
dense_rank() over(order by ordercount desc) as rank_orders,
dense_rank() over(order by overallChange_ordercount asc) as rank_Overall_Drop_Orders
from
(select
dining_segment,
sum(ordercount) as ordercount,
round((sum(ordercount)/sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following))*100,2) as ordercount_share,
round(sum(tot_revenue),2) as tot_revenue,
round((sum(tot_revenue)/sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following))*100,2) as tot_revenue_share
from foodhunter_reports.dining_segment_summary
where day_category = "Week-Day"
group by dining_segment) a
left join
(select dining_segment, round(((lead_ordercount-ordercount)/ordercount)*100,2) as overallChange_ordercount, round(((lead_totRevenue-tot_revenue)/tot_revenue)*100,2) as overallChange_totRev
from
(select year__month, dining_segment,
ordercount, lead(ordercount,3) over(partition by dining_segment order by year__month) as lead_ordercount,
tot_revenue, lead(tot_revenue,3) over(partition by dining_segment order by year__month) as lead_totRevenue
from foodhunter_reports.dining_segment_summary
where day_category = "Week-Day") q1
) q2
on a.dining_segment = q2.dining_segment and q2.overallChange_ordercount is not null and q2.overallChange_totRev is not null)
select
dining_segment, ordercount, ordercount_share, rank_orders, overallChange_ordercount, rank_Overall_Drop_Orders, tot_revenue, tot_revenue_share, rank_Rev, overallChange_totRev, rank_Overall_Drop_Rev, 
round(attentionScore / (select max(attentionScore) from weekday_priority), 2) as attentionScore
from weekday_priority
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('9995'), Decimal('31.82'), 1, Decimal('-21.36'), 4, 275549.9, 31.66, 1, -22.72, 4, 1.0),
 ('4-Snacks', Decimal('5709'), Decimal('18.17'), 2, Decimal('-29.17'), 1, 158525.3, 18.21, 2, -30.65, 1, 0.78),
 ('1-Breakfast', Decimal('4375'), Decimal('13.93'), 3, Decimal('-17.86'), 6, 121183.5, 13.92, 3, -18.02, 6, 0.35),
 ('3-Lunch', Decimal('4326'), Decimal('13.77'), 4, Decimal('-20.77'), 5, 120870.7, 13.89, 4, -19.84, 5, 0.38),
 ('5-Dinner', Decimal('4287'), Decimal('13.65'), 5, Decimal('-27.85'), 3, 119017.2, 13.67, 5, -27.38, 3, 0.52),
 ('2-Brunch', Decimal('2720'), Decimal('8.66'), 6, Decimal('-28.78'), 2, 75276.5, 8.65, 6, -30.34, 2, 0.36)]

#### <b> observations on weekdays </b>
* Overall decline gets intensified for the dining segments: Brunch & Snacks with roughly <b> 30% drop </b> in revenue and order count over the entire four months from June 2025 to September 2025.
* Decline in order count and revenue for the After Hours & Dinner segment is getting slower over the months, but the revenue shows a considerable increase from August to September. This contrast in revenue may be a sign of offering discounts. <i> [need to deepdive] </i>

* The Snacks segment shows an increase in the rate of decline for both order count and revenue over the months.
* The Breakfast, Brunch and Lunch segments have volatility in the rate of drop in both revenue and order count.
* Ordering Dining Segments based on Overall Decline : <b><i> [Snacks] > [Brunch] > [Dinner] > [After hours] > [Lunch] > [BreakFast] </i> </b>
* Ordering Dining Segments based on Total Revenue : <b><i> [After hours] > [Snacks] > [BreakFast] > [Lunch] > [Dinner] > [Brunch] </i> </b>
---
#### <b> weightage assessment </b>
* Order of Focus : <b> <i> [After Hours] > [Snacks] > [Dinner] > [Lunch] > [Brunch] > [Breakfast]
* The After Hours segment shows a highest revenue share and comparitively moderate revenue decline, and it needs high attention and effective adjustments to optimize its operations and outcomes. <b>[P0]</b>

* Snack Segment shows higher decline percentage, high revenue next to the 'After Hours' Segment needs next level attention to prevent further loss. <b>[P1]</b>

In [29]:
%%sql

with weekendlag_dining_segment_summary as (
  select 
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by dining_segment order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by dining_segment order by year__month) as lead_overall_rev
  from  foodhunter_reports.dining_segment_summary 
  where day_category = 'Week-End')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount) / ordercount)*100, 2) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue) / tot_revenue)*100, 2) as totRev_overall_change_in_percent
from weekendlag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/
24 rows affected.


[('2022-6', '1-Breakfast', 485, None, Decimal('-26.80'), 13627.1, None, -21.73),
 ('2022-7', '1-Breakfast', 423, Decimal('-12.78'), None, 11492.2, -15.67, None),
 ('2022-8', '1-Breakfast', 352, Decimal('-16.78'), None, 10024.3, -12.77, None),
 ('2022-9', '1-Breakfast', 355, Decimal('0.85'), None, 10666.5, 6.41, None),
 ('2022-6', '2-Brunch', 322, None, Decimal('-33.85'), 9342.3, None, -36.94),
 ('2022-7', '2-Brunch', 264, Decimal('-18.01'), None, 7157.6, -23.39, None),
 ('2022-8', '2-Brunch', 246, Decimal('-6.82'), None, 7096.9, -0.85, None),
 ('2022-9', '2-Brunch', 213, Decimal('-13.41'), None, 5891.5, -16.98, None),
 ('2022-6', '3-Lunch', 472, None, Decimal('-32.20'), 13442.3, None, -34.67),
 ('2022-7', '3-Lunch', 436, Decimal('-7.63'), None, 12517.6, -6.88, None),
 ('2022-8', '3-Lunch', 396, Decimal('-9.17'), None, 11939.8, -4.62, None),
 ('2022-9', '3-Lunch', 320, Decimal('-19.19'), None, 8782.2, -26.45, None),
 ('2022-6', '4-Snacks', 603, None, Decimal('-30.68'), 16235.5, None, -30.72),
 ('2022-7', '4-Snacks', 559, Decimal('-7.30'), None, 15048.0, -7.31, None),
 ('2022-8', '4-Snacks', 508, Decimal('-9.12'), None, 14322.3, -4.82, None),
 ('2022-9', '4-Snacks', 418, Decimal('-17.72'), None, 11247.4, -21.47, None),
 ('2022-6', '5-Dinner', 467, None, Decimal('-24.20'), 12602.6, None, -19.97),
 ('2022-7', '5-Dinner', 409, Decimal('-12.42'), None, 11082.2, -12.06, None),
 ('2022-8', '5-Dinner', 373, Decimal('-8.80'), None, 9763.9, -11.9, None),
 ('2022-9', '5-Dinner', 354, Decimal('-5.09'), None, 10086.2, 3.3, None),
 ('2022-6', '6-After hours', 1110, None, Decimal('-26.49'), 31670.0, None, -29.58),
 ('2022-7', '6-After hours', 957, Decimal('-13.78'), None, 26310.5, -16.92, None),
 ('2022-8', '6-After hours', 848, Decimal('-11.39'), None, 24632.5, -6.38, None),
 ('2022-9', '6-After hours', 816, Decimal('-3.77'), None, 22301.5, -9.46, None)]

In [30]:
%%sql

#revenue and ordercount shares for dining segments for weekday.
with weekend_priority as(
select
a.dining_segment,
a.ordercount,
a.ordercount_share,
q2.overallChange_ordercount,
a.tot_revenue,
a.tot_revenue_share,
q2.overallChange_totRev,
#just to set priority on framing out recovery plan based on severity
(-(1.5*a.tot_revenue_share) * (q2.overallChange_totRev)) as attentionScore, #providing 0.5 units more priority for revenue share from total revenue decline
dense_rank() over(order by tot_revenue desc) as rank_Rev, 
dense_rank() over(order by overallChange_totRev asc) as rank_Overall_Drop_Rev,
dense_rank() over(order by ordercount desc) as rank_orders,
dense_rank() over(order by overallChange_ordercount asc) as rank_Overall_Drop_Orders
from
(select
dining_segment,
sum(ordercount) as ordercount,
round((sum(ordercount)/sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following))*100,2) as ordercount_share,
round(sum(tot_revenue),2) as tot_revenue,
round((sum(tot_revenue)/sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following))*100,2) as tot_revenue_share
from foodhunter_reports.dining_segment_summary
where day_category = "Week-End"
group by dining_segment) a
left join
(select dining_segment, round(((lead_ordercount-ordercount)/ordercount)*100,2) as overallChange_ordercount, round(((lead_totRevenue-tot_revenue)/tot_revenue)*100,2) as overallChange_totRev
from
(select year__month, dining_segment,
ordercount, lead(ordercount,3) over(partition by dining_segment order by year__month) as lead_ordercount,
tot_revenue, lead(tot_revenue,3) over(partition by dining_segment order by year__month) as lead_totRevenue
from foodhunter_reports.dining_segment_summary
where day_category = "Week-End") q1
) q2
on a.dining_segment = q2.dining_segment and q2.overallChange_ordercount is not null and q2.overallChange_totRev is not null)
select
dining_segment, ordercount, ordercount_share, rank_orders, overallChange_ordercount, rank_Overall_Drop_Orders, tot_revenue, tot_revenue_share, rank_Rev, overallChange_totRev, rank_Overall_Drop_Rev, 
round(attentionScore / (select max(attentionScore) from weekend_priority), 2) as attentionScore
from weekend_priority
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('3731'), Decimal('31.87'), 1, Decimal('-26.49'), 5, 104914.5, 32.06, 1, -29.58, 4, 1.0),
 ('4-Snacks', Decimal('2088'), Decimal('17.84'), 2, Decimal('-30.68'), 3, 56853.2, 17.37, 2, -30.72, 3, 0.56),
 ('3-Lunch', Decimal('1624'), Decimal('13.87'), 3, Decimal('-32.20'), 2, 46681.9, 14.26, 3, -34.67, 2, 0.52),
 ('1-Breakfast', Decimal('1615'), Decimal('13.80'), 4, Decimal('-26.80'), 4, 45810.1, 14.0, 4, -21.73, 5, 0.32),
 ('5-Dinner', Decimal('1603'), Decimal('13.69'), 5, Decimal('-24.20'), 6, 43534.9, 13.3, 5, -19.97, 6, 0.28),
 ('2-Brunch', Decimal('1045'), Decimal('8.93'), 6, Decimal('-33.85'), 1, 29488.3, 9.01, 6, -36.94, 1, 0.35)]

#### <b> observations on weekends</b>
* <b> Weekends show a higher overall decline across dining segments compared to weekdays. </b>
* Overall decline gets intensified for the dining segments: Brunch & Lunch with roughly <b> 35% drop </b> in revenue and ordercount over the entire four months from June 2022 to September 2022.
* During weekends, dining segment - brunch experience fluctuations in the rate of decline in revenue and order count over the four-month period.
* Breakfast shows positive signs from August 2022 to September 2022, as the rate of decline slows down and begins to shift towards rate of elevation in terms of both revenue and order count over the period of time.
* The dinner segment shows positive signs over the months, as the rate of decline in revenue and order count becomes slower over time.
* Lunch and snacks segments show a sharp decline over the months in terms of both order counts and revenue, getting faster as time goes on. It has a higher revenue and order share, making it an important segment to consider for framing recovery strategies.
* The decline in order count for the after-hours segment is slowing, but revenue is falling faster from Aug 2022 to Sep 2022. <i> We need to investigate does marketing strategies like offering discounts causing this. </i>

* Ordering Dining Segments based on Overall Decline : <b><i>[Brunch] > [Lunch] > [Snacks] > [After Hours] > [Breakfast] > [Dinner] </i></b>
* Ordering Dining Segments based on Total Revenue : <b><i> [After Hours] > [Snacks] > [Lunch] > [Breakfast] > [Dinner] > [Brunch]</i></b>
---
#### <b> weightage assessment </b>
* Order of Focus : <b> <i> [After Hours] > [Snacks] > [Lunch] > [Brunch] > [Breakfast] > [Dinner]
  
* After Hours segment shows highest revenue share and also higher decline percentage next to the segments Snacks & Lunch.<b><i> Recommending to treat this as [P0] risk </i></b>
* Snacks and Lunch Segments have higher revenue share and higher decline percentage over the months Jun to Sep. <b> [P1] </b>

### <b> Impact of Discounts on Order Count and Total Revenue </b>

In [52]:
%%sql
select
  month(order_date) as month_no,
  round(avg(discount), 2) as avg_discount,
  round((sum(discount) / sum(final_price))* 100, 3) as discount_ratio_,
  count(distinct order_id) as ordercount,
  round(sum(final_price), 2) as tot_revenue
from foodhunter.orders
group by month(order_date);

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[(6, 3.82, 13.751, 12502, 347577.5),
 (7, 3.74, 13.522, 11144, 308601.5),
 (8, 3.87, 13.813, 10107, 283365.9),
 (9, 3.74, 13.562, 9365, 258161.1)]

#### <b> observation</b>
* By looking at high level, we can say that offering discounts does not influence total revenue & the number of orders.
  
* To confirm this, we should perform causality test between 'avg_discount' and 'ordercount' at week level over the entire timeframe. 

### <b> Avg delivery time & Avg order rating & driver rating</b>

In [91]:
%%sql

create or replace view foodhunter_reports.deliverytime_and_revenue as
  select
    order_id,
    order_date,
    case when dayofweek(order_date) in (1,7) then "Week-End" else "Week-Day" end as day_category,
    customer_id,
    o.driver_id,
    order_time,
    delivered_time,
    time_to_sec(timediff(delivered_time, order_time))/60 as delivery_duration_in_mins,
    final_price,
    delivery_fee,
    order_rating,
    rating as driver_rating
  from foodhunter.orders o left join foodhunter.drivers d on o.driver_id = d.driver_id;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.


[]

In [97]:
%%sql

with month_summary as (
select
q1.month_no,
q1.day_category,
q1.avg_delivery_duration,
q2.median_delivery_duration,
q1.avg_order_rating,
q1.ordercount,
q1.total_revenue
from
(select
  month(order_date) as month_no,
  day_category,
  avg(delivery_duration_in_mins) as avg_delivery_duration,
  avg(order_rating) as avg_order_rating,
  count(distinct order_id) as ordercount,
  sum(final_price) as total_revenue
 from foodhunter_reports.deliverytime_and_revenue
 group by month(order_date), day_category
) q1 inner join
(
with ranking as(
select order_date, day_category, delivery_duration_in_mins, rnk, count(*) over(partition by month(order_date), day_category) as total_rows, max(rnk) over(partition by month(order_date), day_category) as max_rnk
from
(select order_date, day_category, delivery_duration_in_mins, row_number() over (partition by month(order_date), day_category order by delivery_duration_in_mins) as rnk from foodhunter_reports.deliverytime_and_revenue) q
)
select
month(order_date) as month_no,
day_category,
case when total_rows % 2 = 1 then max(case when rnk = (total_rows + 1)/2 then delivery_duration_in_mins end) else avg(case when rnk in(total_rows/2, (total_rows/2)+1) then delivery_duration_in_mins end) end as median_delivery_duration
from ranking
group by month(order_date), day_category
) q2 on q1.month_no = q2.month_no and q1.day_category = q2.day_category
),
month_lagsummary as(
select
month_no,
day_category,
avg_delivery_duration,
lag(avg_delivery_duration) over(partition by day_category order by month_no) as lag_avg_delivery_duration,
median_delivery_duration,
lag(median_delivery_duration) over(partition by day_category order by month_no) as lag_median_delivery_duration,
avg_order_rating,
lag(avg_order_rating) over(partition by day_category order by month_no) as lag_order_rating,
ordercount,
lag(ordercount) over(partition by day_category order by month_no) as lag_ordercount,
total_revenue,
lag(total_revenue) over(partition by day_category order by month_no) as lag_revenue
from
month_summary
)
select
month_no,
day_category,
round(avg_delivery_duration,3) as avg_delivery_duration,
round(((avg_delivery_duration - lag_avg_delivery_duration) / (lag_avg_delivery_duration))*100, 2) as delivery_duration_change_percent_avg,
round(median_delivery_duration,3) as median_delivery_duration,
round(((median_delivery_duration - lag_median_delivery_duration) / (lag_median_delivery_duration))*100, 2) as delivery_duration_change_percent_med,
round(avg_order_rating,2) as avg_order_rating,
round(((avg_order_rating - lag_order_rating) / (lag_order_rating))*100, 2) as order_rating_change_percent,
ordercount,
round(((ordercount - lag_ordercount) / (lag_ordercount))*100, 2) as ordercount_change_percent,
round(total_revenue, 2) as total_revenue,
round(((total_revenue - lag_revenue) / lag_revenue)*100, 2) as total_revenue_change_percent
from
month_lagsummary
order by day_category, month_no;

 * mysql+mysqlconnector://root:***@localhost/
8 rows affected.


[(6, 'Week-Day', Decimal('19.130'), None, Decimal('19.000'), None, 4.22, None, 9043, None, 250657.7, None),
 (7, 'Week-Day', Decimal('21.130'), Decimal('10.45'), Decimal('21.000'), Decimal('10.53'), 3.71, -12.02, 8096, Decimal('-10.47'), 224993.4, -10.24),
 (8, 'Week-Day', Decimal('24.163'), Decimal('14.36'), Decimal('24.000'), Decimal('14.29'), 3.38, -8.95, 7384, Decimal('-8.79'), 205586.2, -8.63),
 (9, 'Week-Day', Decimal('31.173'), Decimal('29.01'), Decimal('31.000'), Decimal('29.17'), 2.99, -11.44, 6889, Decimal('-6.70'), 189185.8, -7.98),
 (6, 'Week-End', Decimal('19.206'), None, Decimal('19.000'), None, 4.22, None, 3459, None, 96919.8, None),
 (7, 'Week-End', Decimal('21.176'), Decimal('10.25'), Decimal('21.000'), Decimal('10.53'), 3.74, -11.39, 3048, Decimal('-11.88'), 83608.1, -13.73),
 (8, 'Week-End', Decimal('24.040'), Decimal('13.53'), Decimal('24.000'), Decimal('14.29'), 3.38, -9.55, 2723, Decimal('-10.66'), 77779.7, -6.97),
 (9, 'Week-End', Decimal('30.968'), Decimal('28.82'), Decimal('31.000'), Decimal('29.17'), 3.01, -10.78, 2476, Decimal('-9.07'), 68975.3, -11.32)]

#### <b> observation</b>
* We observe inverse relationship between delivery duration and order count, tot_revenue as well as order rating. This suggests that delivery duration acts as one of the factors which impact revenue, order count and order rating.
* To confirm this, we should perform causality test between delivery_duration and Order Count/ Total Revenue/ Order Rating.

### <b> Food Items preferences</b>

In [35]:
%%sql

select
fi.food_type, count(distinct o.order_id) as ordercount, sum(oi.quantity) as quantity_sum, count(distinct fi.item_id) as item_count, count(distinct r.cuisine) as cuisine_count
from foodhunter.orders o
join foodhunter.orders_items oi using(order_id)
join foodhunter.food_items fi using(item_id)
join foodhunter.restaurants r using(restaurant_id)
group by fi.food_type;

 * mysql+mysqlconnector://root:***@localhost/
2 rows affected.


[('non-veg', 34664, Decimal('63077'), 219, 4),
 ('veg', 15454, Decimal('23239'), 81, 4)]

In [36]:
%%sql

update foodhunter.food_items 
set food_type = case
                 when food_type like 'v%' then 'veg'
                 when food_type like 'non-%' then 'non-veg'
                 else 'others'
end;

 * mysql+mysqlconnector://root:***@localhost/
314 rows affected.


[]

In [37]:
%%sql

create or replace view foodhunter_reports.food_type_orders as
select
fi.food_type, count(distinct o.order_id) as ordercount, sum(oi.quantity) as quantity_sum, count(distinct fi.item_id) as item_count, count(distinct r.cuisine) as cuisine_count
from foodhunter.orders o
join foodhunter.orders_items oi using(order_id)
join foodhunter.food_items fi using(item_id)
join foodhunter.restaurants r using(restaurant_id)
group by fi.food_type;

select
*
from
foodhunter_reports.food_type_orders;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
2 rows affected.


[('non-veg', 34664, Decimal('63077'), 219, 4),
 ('veg', 15454, Decimal('23239'), 81, 4)]

In [38]:
%%sql
#Items that have not been ordered by the customers over the 4 months from June to September

create or replace view foodhunter_reports.deaditems_for_four_months as
select
item_id, item_name, food_type, restaurant_id, restaurant_name, cuisine, price
from foodhunter.orders_items
right join foodhunter.food_items using(item_id)
join foodhunter.restaurants using(restaurant_id)
where order_id is null;

select
*
from
foodhunter_reports.deaditems_for_four_months;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
14 rows affected.


[(414, 'Pepperoni pizza', 'non-veg', 99, 'Egg Shop', 'Indian', None),
 (412, 'Bellini', 'veg', 58, 'Le Coucou', 'Italian', None),
 (406, 'Special Pasta', 'non-veg', 13, "Roberta's Pizza", 'Italian', None),
 (407, 'Pepperoni pizza', 'non-veg', 13, "Roberta's Pizza", 'Italian', None),
 (408, 'Papadum', 'veg', 16, 'La Vara', 'Italian', None),
 (411, 'Special Pasta', 'non-veg', 58, 'Le Coucou', 'Italian', None),
 (413, 'Special Pasta', 'non-veg', 60, 'Frenchette', 'Italian', None),
 (403, 'Nam Tok', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (405, 'Tod Man Pla', 'non-veg', 75, 'Pam Real Thai', 'Thai', None),
 (409, 'Fiano di Avellino', 'veg', 35, 'Buvette', 'Italian', None),
 (402, 'Kana Moo Krob', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (401, 'Pad Sator', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (404, 'Kana Moo Krob', 'non-veg', 75, 'Pam Real Thai', 'Thai', None),
 (410, 'Cozze al vino bianco', 'non-veg', 35, 'Buvette', 'Italian', None)]

#### <b> observations</b>
* Non-veg items are preferred over veg items.
* We should consider expanding the variety of non-vegetarian food items and introducing a wider range of regional and international non-veg cuisines to attract more customers.
* At the same time, we can adopt a unique marketing strategy for vegetarian dishes by reintroducing traditional and culturally rich recipes — especially those that date back 5–6 centuries, whether from local heritage or other fellow cultures. This approach will create curiosity and appeal for customers looking to try something new, authentic and deeply rooted in history.
* We can observe that most of the dishes not preferred by people over the four months from June to September are Italian (8 out of 14 dishes), followed by Thai dishes (5 out of 14) and then only one Indian dish.
* The following restaurants have not received any orders through our FoodHunter app: <b><i> Egg Shop, Le Coucou, Roberta's Pizza, La Vara, Frenchette, Minetta Tavern, Pam Real Thai and Buvette.</i> </b>
* Non-vegetarian dishes hold a larger share than vegetarian dishes in the not preferred list.

### <b> customer demographics analysis</b>

In [39]:
%%sql
#checking for mismatches in addresses
select
count(*) address_mismatch
from foodhunter.orders o
join foodhunter.customers c on o.customer_id = c.customer_id
where o.delivery_address <> c.address;

 * mysql+mysqlconnector://root:***@localhost/
1 rows affected.


[(0,)]

In [40]:
%%sql
#top 10 -performing customer locations
create or replace view foodhunter_reports.top_10_bestperforming_loc as
select
place_name,
round(sum(final_price), 2) as tot_revenue,
count(distinct order_id) as ordercount,
count(distinct customer_id) as customer_count
from
  (select
     o.order_id,
     o.final_price,
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from foodhunter.orders o
   inner join foodhunter.customers c on o.customer_id = c.customer_id) q1
group by place_name
order by tot_revenue desc, ordercount desc, customer_count desc, place_name asc
limit 10;

select
*
from
foodhunter_reports.top_10_bestperforming_loc;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
10 rows affected.


[(' Parkview Avenue', 10890.4, 343, 73),
 (' Oak Street', 10807.2, 364, 77),
 (' Greenfield Terrace', 10781.1, 368, 77),
 (' Washington Avenue', 10674.9, 354, 74),
 (' Chestnut Avenue', 10491.0, 376, 75),
 (' Beechwood Road', 10430.9, 357, 75),
 (' High Street', 10206.3, 345, 76),
 (' Glenwood Avenue', 10194.9, 333, 76),
 (' Northview Road', 10152.0, 376, 75),
 (' Brookdale Place', 10108.4, 339, 76)]

In [41]:
%%sql
#lowest 10 - performing customer locations 
create or replace view foodhunter_reports.top_10_worstperforming_loc as
select
place_name,
round(sum(final_price), 2) as tot_revenue,
count(distinct order_id) as ordercount,
count(distinct customer_id) as customer_count
from
  (select
     o.order_id,
     o.final_price,
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from foodhunter.orders o
   inner join foodhunter.customers c on o.customer_id = c.customer_id) q1
group by place_name
order by tot_revenue asc, ordercount asc, customer_count asc, place_name asc
limit 10;

select
*
from
foodhunter_reports.top_10_worstperforming_loc;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
10 rows affected.


[(' Belmont Avenue', 7115.1, 262, 73),
 (' Orchard Road', 7267.0, 303, 75),
 (' Holly Lane', 7416.2, 292, 69),
 (' Beech Street', 7437.0, 292, 71),
 (' Yorktown Road', 7502.8, 294, 71),
 (' Northview Terrace', 7601.8, 301, 73),
 (' Brookside Avenue', 7608.2, 285, 73),
 (' Pinehurst Drive', 7766.1, 283, 70),
 (' Valley View Road', 7778.2, 287, 71),
 (' Maplewood Road', 7812.7, 305, 72)]

In [42]:
%%sql
#places with inactive orders over the 4 Months

create or replace view foodhunter_reports.places_with_inactiveorders as
select
place_name,
count(distinct customer_id) as customer_count
from
  (select
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from foodhunter.orders o
   right join foodhunter.customers c on o.customer_id = c.customer_id
   where o.order_id is null) q1
group by place_name
order by customer_count desc, place_name asc;

select
*
from
foodhunter_reports.places_with_inactiveorders
limit 10;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
10 rows affected.


[(' Fairview Avenue', 4),
 (' Oak Hill Drive', 4),
 (' Springdale Road', 4),
 (' Briarwood Court', 3),
 (' Cedar Street', 3),
 (' Elmwood Road', 3),
 (' Grove Street', 3),
 (' Longview Road', 3),
 (' Maplewood Road', 3),
 (' Northview Terrace', 3)]

#### <b> observations</b>
* <b> Top Performing Customer Locations (by revenue, orders, and customer count): </b>
<i> Parkview Avenue > Oak Street > Greenfield Terrace > Washington Avenue > Chestnut Avenue > Beechwood Road > High Street > Glenwood Avenue > Northview Road > Brookdale Place. </i>
*  <b> Lowest Performing Customer Locations (by revenue, orders, and customer count): </b>
<i> Belmont Avenue < Orchard Road < Holly Lane < Beech Street < Yorktown Road < Northview Terrace < Brookside Avenue < Pinehurst Drive < Valley View Road < Maplewood Road. </i>
* <b> Top 10 Places with Inactive orders over the 4 Months (by customer base): </b>
<i> Fairview Avenue > Oak Hill Drive > Springdale Road > Briarwood Court > Cedar Street > Elmwood Road > Grove Street > Longview Road > Maplewood Road > Northview Terrace </i>

### <b> Meal Pricing Analysis </b>

In [43]:
%%sql

create or replace view foodhunter_reports.price_category_ordercount_revenue as
select
price_category,
total_revenue,
dense_rank() over(order by total_revenue desc) as total_revenue_rank,
ordercount,
dense_rank() over(order by ordercount desc) as ordercount_rank,
round(total_revenue / ordercount, 2) as price_per_order
from
  (select
    price_category,
    count(distinct order_id) as ordercount,
    round(sum(final_price), 2) as total_revenue
   from
     (select
       *,
       case when total_price between 3 and 10 then "Class A"
         when total_price between 11 and 25 then "Class B"
         when total_price between 26 and 60 then "Class C"
         when total_price between 61 and 100 then "Class D"
         when total_price between 101 and 150 then "Class E"
       end as price_category
      from foodhunter.orders) q1
   group by price_category) q2
order by price_category;

select
*
from
foodhunter_reports.price_category_ordercount_revenue;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
5 rows affected.


[('Class A', 127744.8, 5, 10913, 3, 11.71),
 ('Class B', 328513.3, 2, 16538, 1, 19.86),
 ('Class C', 387797.1, 1, 11817, 2, 32.82),
 ('Class D', 168298.8, 4, 2162, 4, 77.84),
 ('Class E', 185352.0, 3, 1688, 5, 109.81)]

#### <b> observations</b>
* we get the most money from fewer orders (comparitive to Class B), bigger purchases (Class C).
* we get the most orders from smaller purchases (price per orders) (Class B).
* Formulate strategies for Class C and Class B categories to get the most out of customer spending.